### PR TITLE
[JENKINS-24805] Addressing comments from PR-12

### PR DESCRIPTION
**Pull request by escoem**
Please, @reviewbybees , specially @jglick @idstein 

This PR uses https://github.com/jenkinsci/credentials-binding-plugin/pull/12 and is focused only in https://issues.jenkins-ci.org/browse/JENKINS-24805

### CODE REVIEW

**Comment by escoem** [comment]
fixed @idstein , thanks!

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java [hunk 1]
```diff
@@ -128,16 +130,25 @@
...
+            public int compare(String o1, String o2) {
+                return o2.length() - o1.length();
+            }
+        };
+
         Filter(Collection<String> secrets, String charsetName) {
             StringBuilder b = new StringBuilder();
-            for (String secret : secrets) {
+            List<String> hiddenSecrets = new ArrayList<String>(secrets);
+            Collections.sort(hiddenSecrets, StringLengthComparator);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java]: This seems like it is fixing some unrelated, possibly unfiled bug, or am I missing something?

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/FileBinding.java [hunk 2]
```diff
@@ -49,7 +49,9 @@
     }
 
     @Override public SingleEnvironment bindSingle(Run<?,?> build, FilePath workspace, Launcher launcher, TaskListener listener) throws IOException, InterruptedException {
-        FileCredentials credentials = getCredentials(build);
+        if(workspace==null)
+        	return new SingleEnvironment("");
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/FileBinding.java]: :bug: the `workspace` parameter to `Binding.bindSingle` should be marked `@Nonnull` (ditto `MultiBinding.bind`) and this should be reverted.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java [hunk 3]
```diff
@@ -72,6 +75,25 @@
         };
     }
 
+    @Override public OutputStream decorateLogger(AbstractBuild build, OutputStream logger)
+            throws IOException, InterruptedException, RunnerAbortedException {
+
+        Map<String,String> overrides = new HashMap<String, String>();
+        for (MultiBinding<?> binding : bindings) {
+            MultiBinding.MultiEnvironment environment = binding.bind(build, null, null, null);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java]: :bug: no do not call `bind` here please.

File: src/test/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStepTest.java [hunk 4]
```diff
@@ -262,6 +262,31 @@ private static FilePath tempDir(FilePath ws) {
             }
         });
     }
+    
+    @Issue("JENKINS-24805")
+    @Test public void maskingSimilarSecrets() {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStepTest.java]: Again this seems like some unrelated bug.

Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStepTest.java]: …while there is no test reproducing the _reported_ bug, affecting only freestyle projects.

**Comment by jglick** [comment]
Sorry, but I think I just need to redo these fixes from scratch myself, based on hints from the existing PRs.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java (hunk 3 above)
Comment (escoem) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java]: I will change it.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java (hunk 1 above)
Comment (escoem) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java]: I will prepare a test for that

**Comment by escoem** [comment]
no problem Jesse, I will try hack meanwhile

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java (hunk 1 above)
Comment (escoem) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java]: There is a test for that.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java [hunk 5]
```diff
@@ -72,12 +85,44 @@
...
+            build.addAction(new SharedMultiEnvAction(environment));
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java]: This action is not used anywhere, as far as I can see.

Comment (escoem) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java]: https://github.com/escoem/credentials-binding-plugin/blob/773fa40ab3fed6954080f596e5f8668a65c40d46/src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java#L65-L86

decorateLogger is called (only one time) before setUp

File: src/test/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapperTest.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    @Issue("JENKINS-24805")
+    @Test public void maskingFreeStyleSecrets() throws Exception {
+        String credentialsId_1 = "creds_1";
+        String username_1 = "s3cr3t";
+        String password_1 = "p4ss";
+        StringCredentialsImpl c_1 = new StringCredentialsImpl(CredentialsScope.GLOBAL, credentialsId_1, "sample1", Secret.fromString(password_1));
+        String credentialsId_2 = "creds_2";
+        String username_2 = "s3cr3t0";
+        String password_2 = "p4ss" + "EvenLonger";
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapperTest.java]: Again this seems to be an unrelated bug (affecting Pipeline equally) that would best be filed and fixed in a separate PR.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java (hunk 3 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java]: No, this is just no good. `Run.createBuildListener` calls `decorateLogger` before `BuildWrapper.setUp` is run. So we simply cannot use the `bind` method as currently designed. It needs an API change in `MultiBinding`.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java [hunk 7]
```diff
@@ -53,10 +63,13 @@
...
 
     @Override public Environment setUp(AbstractBuild build, final Launcher launcher, BuildListener listener) throws IOException, InterruptedException {
-        final List<MultiBinding.MultiEnvironment> m = new ArrayList<MultiBinding.MultiEnvironment>();
-        for (MultiBinding binding : bindings) {
-            m.add(binding.bind(build, build.getWorkspace(), launcher, listener));
-        }
+        final List<MultiBinding.MultiEnvironment> m = Lists.transform(build.getActions(SharedMultiEnvAction.class),
+                new Function<SharedMultiEnvAction, MultiBinding.MultiEnvironment>() {
+                    public MultiBinding.MultiEnvironment apply (SharedMultiEnvAction action) {
+                        return action.get();
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java]: :bug: There is no way this can work. You probably do not realize it because there are no tests for `FileBinding` or `ZipFileBinding` with `SecretBuildWrapper`.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java [hunk 8]
```diff
@@ -72,12 +85,44 @@
...
+    }
+
     @Override public void makeSensitiveBuildVariables(AbstractBuild build, Set<String> sensitiveVariables) {
         for (MultiBinding binding : bindings) {
             sensitiveVariables.addAll(binding.variables());
         }
     }
 
+    private static class SharedMultiEnvAction extends InvisibleAction {
+        private final transient MultiBinding.MultiEnvironment me;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java]: Adding an `Action` with only a `transient` field is pointless. Better to just keep a `WeakHashMap`.

**Comment by stephenc** [comment]
@jglick what is the state of this PR? Is it closed to be superseded by a PR from you or are you expecting changes from the originator.

**Comment by aheritier** [comment]
@stephenc Based on @jglick comment on July 7th I think it should be closed

> Sorry, but I think I just need to redo these fixes from scratch myself, based on hints from the existing PRs.
